concept/thing/thing_manager.rs

```diff
@@ -806,31 +838,226 @@ impl ThingManager {
+                            owner_type.get_label(snapshot, self.type_manager())?.clone(),
+                            attribute_type.get_label(snapshot, self.type_manager())?.clone(),
+                        )),
+                        Some(owns) => Ok(owns),
+                    }?
+                    .into_owned();
+
+                self.add_exclusive_lock_for_owns_cardinality_constraint(snapshot, &object, owns.clone())?;
+                self.add_exclusive_lock_for_unique_constraint(snapshot, &object, attribute_value, owns)?;
+            } else if ThingEdgeLinks::is_links(key_reference) {
+                let role_player = ThingEdgeLinks::new(Bytes::Reference(key_reference.byte_ref()));
+                let relation = Relation::new(role_player.relation()).into_owned();
+                let player = Object::new(role_player.player()).into_owned();
+                let relation_type = relation.type_();
+                let owner_type = player.type_();
+                let role_type = RoleType::build_from_type_id(role_player.role_id()).into_owned();
+
+                let plays = match owner_type.get_plays_role(snapshot, self.type_manager(), role_type.clone())? {
+                    None => Err(ConceptReadError::CannotGetPlaysDoesntExist(
+                        owner_type.get_label(snapshot, self.type_manager())?.clone(),
+                        role_type.get_label(snapshot, self.type_manager())?.clone(),
+                    )),
+                    Some(plays) => Ok(plays),
+                }?
+                .into_owned();
+                self.add_exclusive_lock_for_plays_cardinality_constraint(snapshot, &player, plays)?;
+
+                let relates = match relation_type.get_relates_role(snapshot, self.type_manager(), role_type.clone())? {
+                    None => Err(ConceptReadError::CannotGetPlaysDoesntExist(
+                        owner_type.get_label(snapshot, self.type_manager())?.clone(),
+                        role_type.get_label(snapshot, self.type_manager())?.clone(),
+                    )),
+                    Some(plays) => Ok(plays),
+                }?
+                .into_owned();
+                self.add_exclusive_lock_for_relates_cardinality_constraint(snapshot, &relation, relates)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn add_exclusive_lock_for_unique_constraint<'a>(
+        &self,
+        snapshot: &mut impl WritableSnapshot,
+        owner: &Object<'a>,
+        value: Value<'a>,
+        owns: Owns<'static>,
+    ) -> Result<(), ConceptReadError> {
+        if let Some(uniqueness_source) = get_uniqueness_source(snapshot, self, owns)? {
+            let lock_key = create_custom_lock_key(
+                [
+                    &Infix::PropertyAnnotationUnique.infix_id().bytes(),
+                    uniqueness_source.attribute().vertex().bytes().bytes(),
+                    value.encode_bytes::<128>().bytes(), // TODO: Where to get 128 from?
```

Comment:
Do we want to encode the whole value here? If yes, I feel like it'd be cool to just place this `128` in `value.rs` as `SOMETHING_DEFAULT` as we already reference it in multiple places.

Reply:
@flyingsilverfin What do you think?

Reply:
You could make a constant in `//resources` which is something like `AD_HOC_INLINE_BYTES` and set that to `128`. This is basically the size of the array allocated for short-lived encode/decode. Better if we don't go to the heap much for these unless they're very large.

Reply:
Done, thx!